signer/kms/gcp/client.go

```diff
@@ -0,0 +1,399 @@
+	}
+
+	if ctx == nil {
+		ctx = context.Background()
+	}
+
+	g := &gcpClient{
+		kvCache: nil,
+	}
+
+	var err error
+	g.projectID, g.locationID, g.keyRing, g.keyName, g.version, err = parseReference(ksp.Reference)
+	if err != nil {
+		return nil, err
+	}
+
+	// NOTE: We need to think about how we want to handle passing in options here
+	g.client, err = gcpkms.NewKeyManagementClient(ctx)
+	if err != nil {
+		return nil, fmt.Errorf("new gcp kms client: %w", err)
+	}
+
+	g.kvCache = ttlcache.New[string, cryptoKeyVersion](
+		ttlcache.WithDisableTouchOnHit[string, cryptoKeyVersion](),
+	)
+
+	// prime the cache
+	g.kvCache.Get(cacheKey)
+	return g, nil
+}
+
+type cryptoKeyVersion struct {
+	CryptoKeyVersion *kmspb.CryptoKeyVersion
+	Verifier         cryptoutil.Verifier
+	PublicKey        crypto.PublicKey
+	HashFunc         crypto.Hash
+}
+
+// use a consistent key for cache lookups
+const cacheKey = "crypto_key_version"
+
+func (g *gcpClient) Verifier() (cryptoutil.Verifier, error) {
+	crv, err := g.getCKV()
+	if err != nil {
+		return nil, fmt.Errorf("transient error while getting KMS verifier: %w", err)
+	}
+
+	return crv.Verifier, nil
+}
+
+// keyVersionName returns the first key version found for a key in KMS
+func (g *gcpClient) keyVersionName(ctx context.Context) (*cryptoKeyVersion, error) {
+	parent := fmt.Sprintf("projects/%s/locations/%s/keyRings/%s/cryptoKeys/%s", g.projectID, g.locationID, g.keyRing, g.keyName)
+
+	parentReq := &kmspb.GetCryptoKeyRequest{
+		Name: parent,
+	}
+	key, err := g.client.GetCryptoKey(ctx, parentReq)
+	if err != nil {
+		return nil, err
+	}
+	if key.Purpose != kmspb.CryptoKey_ASYMMETRIC_SIGN {
+		return nil, errors.New("specified key cannot be used to sign")
+	}
+
+	// if g.version was specified, use it explicitly
+	var kv *kmspb.CryptoKeyVersion
+	if g.version != "" {
+		req := &kmspb.GetCryptoKeyVersionRequest{
+			Name: parent + fmt.Sprintf("/cryptoKeyVersions/%s", g.version),
+		}
+		kv, err = g.client.GetCryptoKeyVersion(ctx, req)
+		if err != nil {
+			return nil, err
+		}
+	} else {
+		req := &kmspb.ListCryptoKeyVersionsRequest{
+			Parent:  parent,
+			Filter:  "state=ENABLED",
+			OrderBy: "name desc",
+		}
+		iterator := g.client.ListCryptoKeyVersions(ctx, req)
+
+		// pick the key version that is enabled with the greatest version value
+		kv, err = iterator.Next()
+		if err != nil {
+			return nil, fmt.Errorf("unable to find an enabled key version in GCP KMS: %w", err)
+		}
+	}
+
+	pubKey, err := g.fetchPublicKey(ctx, kv.Name)
+	if err != nil {
+		return nil, fmt.Errorf("unable to fetch public key while creating signer: %w", err)
+	}
+
+	// kv is keyVersion to use
+	crv := cryptoKeyVersion{
+		CryptoKeyVersion: kv,
+		PublicKey:        pubKey,
+	}
+
+	// crv.Verifier is set here to enable storing the public key & hash algorithm together,
+	// as well as using the in memory Verifier to perform the verify operations.
+	switch kv.Algorithm {
+	case kmspb.CryptoKeyVersion_EC_SIGN_P256_SHA256:
```

Comment:
similar to my comment on the AWS verifier, I think some of this can be simplified to just picking up which hash to use and calling `cryptoutil.NewVerifier` with the `VerifyWithHash` option

Reply:
yeah okay sure, I've modified and tested for AWS, just need to do the same for GCP.